metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,455 @@
+        });
+    }
+}
+
+void DBGSSHash::call_incoming_kmers(node_index node,
+                                    const IncomingEdgeCallback& callback) const {
+    if (mode_ == CANONICAL) {
+        call_incoming_kmers_with_rc<true>(node, [&](node_index prev, char c, bool orientation) {
+            callback(orientation ? reverse_complement(prev) : prev, c);
+        });
+    } else {
+        call_incoming_kmers_with_rc<false>(node, [&](node_index prev, char c, bool) {
+            callback(prev, c);
+        });
+    }
+}
+
+size_t DBGSSHash::outdegree(node_index node) const {
+    assert(node > 0 && node <= num_nodes());
+    size_t res = 0;
+    adjacent_outgoing_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+size_t DBGSSHash::indegree(node_index node) const {
+    assert(node > 0 && node <= num_nodes());
+    size_t res = 0;
+    adjacent_incoming_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+void DBGSSHash::call_nodes(
+        const std::function<void(node_index)>& callback,
+        const std::function<bool()> &terminate) const {
+    for (size_t node_idx = 1; !terminate() && node_idx <= dict_size(); ++node_idx) {
+        callback(node_idx);
+    }
+
+    if (mode_ == CANONICAL) {
+        for (size_t node_idx = 1; !terminate() && node_idx <= dict_size(); ++node_idx) {
+            size_t rc_node_idx = reverse_complement(node_idx);
+            if (rc_node_idx != node_idx)
+                callback(rc_node_idx);
+        }
+    }
+}
+
+template <bool with_rc>
+std::pair<DBGSSHash::node_index, bool>
+DBGSSHash::kmer_to_node_with_rc(std::string_view kmer) const {
+    if (!num_nodes())
+        return std::make_pair(npos, false);
+
+    return std::visit([&](const auto &d) {
+        auto res = d.lookup_advanced(kmer.data(), with_rc);
+        return std::make_pair(sshash_to_graph_index(res.kmer_id), res.kmer_orientation);
+    }, dict_);
+}
+template
+std::pair<DBGSSHash::node_index, bool>
+DBGSSHash::kmer_to_node_with_rc<true>(std::string_view) const;
+template
+std::pair<DBGSSHash::node_index, bool>
+DBGSSHash::kmer_to_node_with_rc<false>(std::string_view) const;
+
+DBGSSHash::node_index DBGSSHash::kmer_to_node(std::string_view kmer) const {
+    if (mode_ == CANONICAL) {
+        auto res = kmer_to_node_with_rc<true>(kmer);
+        return res.second ? reverse_complement(res.first) : res.first;
+    } else {
+        return kmer_to_node_with_rc<false>(kmer).first;
+    }
+}
+
+std::string DBGSSHash::get_node_sequence(node_index node) const {
+    assert(node > 0 && node <= num_nodes());
+    std::string str_kmer(k_, ' ');
+    node_index node_canonical = node > dict_size() ? node - dict_size() : node;
+    uint64_t ssh_idx = graph_index_to_sshash(node_canonical);
+    std::visit([&](const auto &d) { d.access(ssh_idx, str_kmer.data()); }, dict_);
+
+    if (node > node_canonical)
+        ::reverse_complement(str_kmer.begin(), str_kmer.end());
+
+    return str_kmer;
+}
+
+
+struct generic_loader {
```

Comment:
Should we try merging this into loader/saver upstream, rather than doing it here? Or at least do it in our fork of essentials and add it as a submodule directly, rather than using it as a submodule of a submodule of a submodule.

Reply:
if we fork essentials and add this, won't it cause problems with `pthash` since there'll be two different versions of the same methods? Also, `dictionary::visit` is designed to work with anything that has interface (and we only use it once here), so unless Giulio has a need for a generic loader, we can just keep it here

Reply:
I've made a bunch of forks for essentials and pthash and propagated necessary changes through them. It also includes making saving const, so in any case I think this is the way we should use them. I will also try to merge these changes upstream.